### PR TITLE
Reconcile multiple PostgreSQL users and databases

### PR DESCRIPTION
**Pull request by cbandy**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?

**Type of Changes:**

 
 - [x] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [x] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**

We create one PostgreSQL user and database that match the name of the PostgresCluster. Credentials and connection details are stored in the `{cluster}-pguser` Secret, but the password cannot be rotated. This cannot be disabled.

**What is the new behavior (if this is a feature change)?**

This change allows one to specify any number of PostgreSQL users that can connect to any number of databases. We reconcile the users' role attributes and passwords, allowing one to grant superuser access or manually trigger credential rotations.

By default, we still create a user and database matching the cluster name, but only when the cluster name is also a valid user name. One can disable this entirely by specifying an empty user list.

We store the password of each user and their connection details in a separate Secret (per user). In addition to the `postgres-operator.crunchydata.com/pguser={username}` label from before, each Secret now has a `postgres-operator.crunchydata.com/role=pguser` label and the Secret name includes the username.

The default Secret has moved to conform to this new scheme, but an old copy remains in place so one can upgrade. Where one referenced `{cluster}-pguser` in the past, they should now reference `{cluster}-pguser-{cluster}`. Once all references are updated, the old copy can be deleted manually.

This also lifts an artificial restriction we had that disallowed clusters named `postgres`. One may now include `postgres` in the user list to generate a Secret and password and allow it to connect over the network.

**Other information**:

Issue: [ch11512]
Issue: [ch11806]
Fixes: #1925
Fixes: #2542
Closes: #2579

### CODE REVIEW

**Review by jmckulk — COMMENTED** [review]
Didn't finish reviewing but its 5:50 on Friday. Here is what I have so far. There were some names that I had a hard time following but I'm not sure they are worth changing. Maybe that would be something to chat about?

**Comment by jmckulk** [comment]
To finish the thought on the naming comment. I feel like I read the word user a lot and in a few different contexts. I'm wondering if there is a way to simplify/clarify. There may not be

File: internal/naming/selectors.go [hunk 1]
```diff
@@ -88,17 +88,23 @@ func ClusterPatronis(cluster *v1beta1.PostgresCluster) metav1.LabelSelector {
...
 
+// ClusterPostgresUsers selects things labeled for PostgreSQL users in cluster.
+func ClusterPostgresUsers(cluster string) metav1.LabelSelector {
+	return metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			LabelCluster: cluster,
+		},
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			// The now-deprecated default PostgreSQL user secret lacks a LabelRole.
+			// The existence of a LabelPostgresUser matches it and current secrets.
```
Comment (jmckulk) [inline internal/naming/selectors.go:99]: > The existence of a LabelPostgresUser matches it and current secrets.

:question: I'm not sure what this comment is trying to say. Should this selector be able to grab a deprecated user secret? iiuc, LabelPostgresUser will only work for new secrets

File: internal/postgres/exec.go [hunk 2]
```diff
@@ -66,3 +66,27 @@ done <<< "${databases}"
 		append([]string{"bash", "-ceu", "--", script, "-"}, args...)...)
 	return stdout.String(), stderr.String(), err
 }
+
+// Execute uses "psql" to execute sql. The sql statement(s) are passed via stdin
+// and may contain psql variables that are assigned from the variables map.
+// - https://www.postgresql.org/docs/current/app-psql.html#APP-PSQL-VARIABLES
+func (exec Executor) Execute(
```
[the remark this answers is not in the document]
Comment (cbandy) [reply internal/postgres/exec.go:73]: Matching sounds good. I'll go with `Exec`.

File: internal/controller/postgrescluster/postgres.go [hunk 3]
```diff
@@ -18,18 +18,441 @@ package postgrescluster
...
+func (r *Reconciler) reconcilePostgresDatabases(
+	ctx context.Context, cluster *v1beta1.PostgresCluster, instances *observedInstances,
+) error {
+	const container = naming.ContainerDatabase
+	var podExecutor postgres.Executor
+
+	// Find the PostgreSQL instance that can execute SQL that writes system
+	// catalogs. When there is none, return early.
+
+	for _, instance := range instances.forCluster {
```
[the remark this answers is not in the document]
Comment (cbandy) [reply internal/controller/postgrescluster/postgres.go:149]: > 🌱 Seems like a check for "can i run sql and where" could be pretty common. Maybe a place for common or util function?

Agreed. The current uses seem to be for writing. 🤔 Reading could involve replicas, but it's safer to interact with the primary.

> 💭 maybe move the check into the exec (assuming thats the only place we need to know the primary/writable instance) and have that return an error or something if it didn't work or couldn't find the right instance.

This method returns `nil` when a good candidate cannot be found. It's not fatal that a cluster lacks a primary, so we shouldn't treat it as an error. The extreme case is a standby cluster; it is never writable.

File: internal/naming/selectors.go (hunk 1 above)
Comment (cbandy) [reply internal/naming/selectors.go:99]: I renamed the constant from `LabelUserSecret` to `LabelPostgresUser`. Both old and new have it; the value is the Postgres user name.

Comment (jmckulk) [reply internal/naming/selectors.go:99]: Ah, I see. Sorry didn't catch that

File: docs/content/architecture/user-management.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
+---
+title: "User Management"
+date:
+draft: false
+weight: 125
+---
+
+PGO manages PostgreSQL users that you define in [`PostgresCluster.spec.users`]({{< relref . "/references/crd#postgresclusterspecusers" >}}).
```
Comment (cbandy) [inline docs/content/architecture/user-management.md:8]: 🔧 This link is wrong.

```suggestion
PGO manages PostgreSQL users that you define in [`PostgresCluster.spec.users`]({{< relref "/references/crd#postgresclusterspecusersindex" >}}).
```

File: docs/content/architecture/user-management.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+[`DROP OWNED`](https://www.postgresql.org/docs/current/sql-drop-owned.html) and
+`DROP USER` or [`DROP ROLE`](https://www.postgresql.org/docs/current/sql-droprole.html)
+in PostgreSQL.
+
+Similarly, removing a database from `PostgresCluster.spec.users` does not drop
+that database nor revoke any access to it. To completely remove it, you must run
+[`DROP DATABASE`](https://www.postgresql.org/docs/current/sql-dropdatabase.html)
+as a superuser in PostgreSQL.
+
+The built-in `postgres` superuser can be managed the same way, which one restriction:
```
Comment (jmckulk) [inline docs/content/architecture/user-management.md:28]:
```suggestion
The built-in `postgres` superuser can be managed the same way, with one restriction:
```

possibly?

**Review by jmckulk — APPROVED** [review]

**Pull request merged**